subsys/pm/pm_shell.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2023 FTP Technologies
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/pm/pm.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/policy.h>
+#include <stdlib.h>
+
+LOG_MODULE_REGISTER(pm_shell);
+
+#define PM_DEVICE_SUSPEND_HELP                                                                     \
+	"Suspend 1 or more devices. Syntax:\n"                                                     \
+	"<device 0> [device 1] .. [device N]"
+
+#define PM_DEVICE_RESUME_HELP                                                                      \
+	"Resume 1 or more devices. Syntax:\n"                                                      \
+	"<device 0> [device 1] .. [device N]"
+
+static int cmd_suspend_devices(const struct shell *sh, size_t argc, char *argv[])
+{
+	const struct device *dev;
+	int i, ret;
+
+	for (i = 1; i < argc; i++) {
+		dev = device_get_binding(argv[i]);
+		if (dev == NULL) {
+			shell_error(sh, "device %s unknown", argv[i]);
+			continue;
+		}
+		if (!device_is_ready(dev)) {
+			shell_error(sh, "device %s not ready", argv[1]);
+			continue;
+		}
+		ret = pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
```

Comment:
-1 on propagating usage of such a dangerous API. Device PM should all go through runtime PM nowadays.

Reply:
I haven't noticed that, I'll check it out

Reply:
This API runs device PM hooks with some minimal guarantees, it's really there for legacy reasons. But as soon as e.g. you have > 1 client on the same device, this API can't be used.

Reply:
The issue you are concerned about is this shell's suspend command could be requested on a device that's currently in use by other code in a system.

The use cases I imagine for the device suspend/resume commands in this shell are:
 - an easy way to test a driver's power management features when developing Zephyr drivers
 - an easy way for Zephyr user's to experiment with sensor/device power draw when optimising a system's power use
 - and the general ability to turn on/off a sensor from the command line for sensor/device experimentation

This is similar to a lot of other Zephyr shells that allow you control/experiment with devices that could be in use by other code in the system. These other shells are also op in and should be used with the understanding that you may be disrupting running code.

If the `pm_device_action_run()` API was to become depreciated or removed the pm shell's suspend/resume commands could be removed.

Reply:
The use case is valid, the way that is been done that is the problem. You should use device runtime this. So, for suspend a device ->`pm_device_runtime_put`, to resume -> `pm_device_runtime_get`. This API (among other things) is referenced count, also serialize pm operations and has other sanity checks. `pm_device_action_run` is intended to be used by the kernel.